### PR TITLE
[ShapeOpt] Introduced scaling of response functions

### PR DESCRIPTION
**Pull request by dbaumgaertner**
This PR:

- provides a generic option to scale all response functions. The scaling is realized completely within the communicator and part of the translation to the standard format.

- cleans the communicator a bit

- slightly improves the check against default values in the optimizer_factory, so that by default, the scaling factor is 1.

### CODE REVIEW

File: applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py [hunk 1]
```diff
@@ -171,33 +177,43 @@ def __deleteAllReportedValues(self):
...
     # --------------------------------------------------------------------------
     def __setValueAsReference(self, response_id, value):
         self.list_of_responses[response_id]["reference_value"] = value
 
     # --------------------------------------------------------------------------
     def __translateValueToStandardForm(self, response_id, value):
-        response = self.list_of_responses[response_id]
-        response_type = response["type"]
+        response_type = self.list_of_responses[response_id]["type"]
+        factor = self.list_of_responses[response_id]["scaling_factor"]
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py]: `factor` -> `scaling_factor`

File: applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py [hunk 2]
```diff
@@ -171,33 +177,43 @@ def __deleteAllReportedValues(self):
...
-                return (response["reference_value"]-value)
+                return factor*(reference_value-value)
             else:
-                return (value-response["reference_value"])
+                return factor*(value-reference_value)
 
     # --------------------------------------------------------------------------
     def __translateGradientToStandardForm(self, response_id, gradient):
         response_type = self.list_of_responses[response_id]["type"]
+        factor = self.list_of_responses[response_id]["scaling_factor"]
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py]: scaling_factor

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py [hunk 3]
```diff
@@ -42,16 +63,35 @@ def CreateOptimizer(optimization_settings, model, external_analyzer=EmptyAnalyze
...
-    communicator = communicator_factory.CreateCommunicator(optimization_settings)
-
-    if optimization_settings["design_variables"]["type"].GetString() == "vertex_morphing":
-        return VertexMorphingMethod(optimization_settings, model_part_controller, analyzer, communicator)
-    else:
-        raise NameError("The following type of design variables is not supported by the optimizer: " + variable_type)
+# ------------------------------------------------------------------------------
+def ValidateObjectives(objectives):
+    for itr in range(objectives.size()):
+        default_settings = Parameters("""
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py]: i would define the defaults outside the loop.

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py [hunk 4]
```diff
@@ -42,16 +63,35 @@ def CreateOptimizer(optimization_settings, model, external_analyzer=EmptyAnalyze
...
-    analyzer = analyzer_factory.CreateAnalyzer(optimization_settings, model_part_controller, external_analyzer)
-
-    communicator = communicator_factory.CreateCommunicator(optimization_settings)
-
-    if optimization_settings["design_variables"]["type"].GetString() == "vertex_morphing":
-        return VertexMorphingMethod(optimization_settings, model_part_controller, analyzer, communicator)
-    else:
-        raise NameError("The following type of design variables is not supported by the optimizer: " + variable_type)
+# ------------------------------------------------------------------------------
+def ValidateObjectives(objectives):
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py]: `objectives` -> `objective_parameters`

The term response, objective, constraint is used for many different types of objects, e.g. the parameters from the json, the dict in the communicator...

**Review by armingeiser — CHANGES_REQUESTED** [review]

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py (hunk 4 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py]: You are right, I changed it here. At some point, we should also make this more consistent throughout the entire app.

**Review by armingeiser — APPROVED** [review]

**Pull request merged**